File: go/custody_client.go
package main

import "fmt"
import "time"
import "strings"
import "sort"
import "net/url"
import "io/ioutil"
import "encoding/hex"
import "crypto/sha256"
import "github.com/btcsuite/btcd/btcec"
import "net/http"

const API_KEY = "x"
const API_SECRET = "x"
const HOST = "https://api.sandbox.cobo.com"
const COBO_PUB = "032f45930f652d72e0c90f71869dfe9af7d713b1f67dc2f7cb51f9572778b9c876"

func SortParams(params map[string]string) string {
	keys := make([]string, len(params))
	i := 0
	for k, _ := range params {
		keys[i] = k
		i++
	}
	sort.Strings(keys)
	sorted := make([]string, len(params))
	i = 0
	for _, k := range keys {
		sorted[i] = k + "=" + url.QueryEscape(params[k])
		i++
	}
	return strings.Join(sorted, "&")
}
func Hash256(s string) string {
	hash_result := sha256.Sum256([]byte(s))
	hash_string := string(hash_result[:])
	return hash_string
}
func Hash256x2(s string) string {
	return Hash256(Hash256(s))
}
func SignEcc(message string) string {
	api_secret, _ := hex.DecodeString(API_SECRET)
	privKey, _ := btcec.PrivKeyFromBytes(btcec.S256(), api_secret)
	sig, _ := privKey.Sign([]byte(Hash256x2(message)))
	return fmt.Sprintf("%x", sig.Serialize())
}

func VerifyEcc(message string, signature string) bool {
	pub_key, _ := hex.DecodeString(COBO_PUB)
	pubKey, _ := btcec.ParsePubKey(pub_key, btcec.S256())

	sigBytes, _ := hex.DecodeString(signature)
	sigObj, _ := btcec.ParseSignature(sigBytes, btcec.S256())

	verified := sigObj.Verify([]byte(Hash256x2(message)), pubKey)
	return verified
}

func Request(method string, path string, params map[string]string) string {
	client := &http.Client{}
	nonce := fmt.Sprintf("%d", time.Now().Unix()*1000)
	sorted := SortParams(params)
	var req *http.Request
	if method == "POST" {
		req, _ = http.NewRequest(method, HOST+path, strings.NewReader(sorted))
		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	} else {
		req, _ = http.NewRequest(method, HOST+path+"?"+sorted, strings.NewReader(""))
	}
	content := strings.Join([]string{method, path, nonce, sorted}, "|")

	req.Header.Set("Biz-Api-Key", API_KEY)
	req.Header.Set("Biz-Api-Nonce", nonce)
	req.Header.Set("Biz-Api-Signature", SignEcc(content))

	resp, _ := client.Do(req)

	defer resp.Body.Close()

	body, _ := ioutil.ReadAll(resp.Body)

	timestamp := resp.Header["Biz-Timestamp"][0]
	signature := resp.Header["Biz-Resp-Signature"][0]
	success := VerifyEcc(string(body)+"|"+timestamp, signature)
	fmt.Println("verify success?", success)
	return string(body)
}

func main() {
	res := Request("GET", "/v1/custody/org_info/", map[string]string{})
	fmt.Println("res", res)
	res = Request("GET", "/v1/custody/transaction_history/", map[string]string{
		"coin": "ETH",
		"side": "deposit",
	})
	fmt.Println("res", res)
}

File: js/custody_client.js
//var bitcoin = require('bitcoinjs-lib') // v3.x.x
//var keyPair = bitcoin.ECPair.fromWIF('L3oe5Wz3wmbxq9tar8nS5bqqCfjMZJNKPXAbbemoHWKXR1SAN8j2')
//var privateKey = keyPair.privateKey
/* global require Buffer*/
const sha256 = require('sha256');
const bip66 = require('bip66');
const fetch = require('node-fetch');
const ec = new require('elliptic').ec('secp256k1')

const { URLSearchParams } = require('url');

const ZERO = Buffer.alloc(1, 0)
function toDER(x){
    let i = 0
    while (x[i] === 0) ++i
    if (i === x.length) return ZERO
    x = x.slice(i)
    if (x[0] & 0x80) return Buffer.concat([ZERO, x], 1 + x.length)
    return x
}

const generate = () => {
    let key = ec.genKeyPair()
    return [key.getPublic(true, "hex"), key.getPrivate("hex")]
}


const sign = (message, api_hex) =>{
    //let message = 'GET|/v1/custody/org_info/|1541560385699|'
    let privateKey = Buffer.from(api_hex, 'hex')
    let result = ec.sign(Buffer.from(sha256.x2(message), 'hex'), privateKey)
    var r = new Buffer(result.r.toString(16, 64), 'hex')
    var s = new Buffer(result.s.toString(16, 64), 'hex')
    r = toDER(r);
    s = toDER(s);
    return bip66.encode(r, s).toString('hex');
};

const coboFetch = (method, path, params, api_key, api_hex, base = 'https://api.sandbox.cobo.com') => {
    let nonce = String(new Date().getTime());
    let sort_params = Object.keys(params).sort().map((k) => {
        return k + '=' + encodeURIComponent(params[k]).replace(/%20/g, "+");
    }).join('&');
    let content = [method, path, nonce, sort_params].join('|');
    let headers = {
        'Biz-Api-Key': api_key,
        'Biz-Api-Nonce': nonce,
        'Biz-Api-Signature': sign(content, api_hex)
    };
    if (method == 'GET'){
        return fetch(base + path + '?' + sort_params, {
            'method': method,
            'headers': headers,
        });
    }else if (method == 'POST'){

        let urlParams = new URLSearchParams();
        for (let k in params){
            urlParams.append(k, params[k])
        }

        return fetch(base + path, {
            'method': method,
            'headers': headers,
            'body': urlParams
        });
    }else{
        throw "unexpected method " + method;
    }
}


let api_key = 'x'
let api_hex = 'x'

coboFetch('GET', '/v1/custody/org_info/', {}, api_key, api_hex)
    .then(res => {
        console.log(res.status);
        res.json().then((data)=>{
            console.log(data);
        })
    }).catch(err => {
        console.log(err)
    });
coboFetch('GET', '/v1/custody/transaction_history/', {'coin': 'ETH', 'side': 'deposit'}, api_key, api_hex)
    .then(res => {
        console.log(res.status);
        res.json().then((data)=>{
            console.log(data);
        })
    }).catch(err => {
        console.log(err)
    });
coboFetch('POST', '/v1/custody/new_address/', {"coin": "ETH"}, api_key, api_hex)
    .then(res => {
        console.log(res.status);
        res.json().then((data)=>{
            console.log(data);
        })
    }).catch(err => {
        console.log(err)
    });
